token-account-based/chaincode-go/chaincode/token_contract.go

```diff
@@ -0,0 +1,173 @@
+package chaincode
+
+import (
+	"fmt"
+	"log"
+	"strconv"
+
+	"github.com/hyperledger/fabric-contract-api-go/contractapi"
+)
+
+// SmartContract provides functions for transferring tokens between accounts
+type SmartContract struct {
+	contractapi.Contract
+}
+
+// Mint creates new tokens and adds them to minter's account balance
+func (s *SmartContract) Mint(ctx contractapi.TransactionContextInterface, amount int) error {
+
+	// Check minter authorization - this sample assumes Org1 is the central banker with privilege to mint new tokens
+	clientMSPID, err := ctx.GetClientIdentity().GetMSPID()
+	if err != nil {
+		return fmt.Errorf("failed to get MSPID: %v", err)
+	}
+	if clientMSPID != "Org1MSP" {
+		return fmt.Errorf("client is not authorized to mint new tokens")
+	}
+
+	// Get ID of submitting client identity
+	minter, err := ctx.GetClientIdentity().GetID()
+	if err != nil {
+		return fmt.Errorf("failed to get client id: %v", err)
+	}
+
+	if amount <= 0 {
+		return fmt.Errorf("mint amount must be a positive integer")
+	}
+
+	currentBalanceBytes, err := ctx.GetStub().GetState(minter)
+	if err != nil {
+		return fmt.Errorf("failed to read minter account %s from world state: %v", minter, err)
+	}
+
+	var currentBalance int
+
+	// If minter current balance doesn't yet exist, we'll create it with a current balance of 0
+	if currentBalanceBytes == nil {
+		currentBalance = 0
+	} else {
+		currentBalance, _ = strconv.Atoi(string(currentBalanceBytes))
+	}
+
+	updatedBalance := currentBalance + amount
+
+	err = ctx.GetStub().PutState(minter, []byte(strconv.Itoa(updatedBalance)))
+	if err != nil {
+		return err
+	}
+
+	log.Printf("minter account %s balance updated from %d to %d", minter, currentBalance, updatedBalance)
+
+	return nil
+}
+
+// Transfer transfers tokens from client account to recipient account.
+// recipient account must be a valid clientID as returned by the ClientID() function.
+func (s *SmartContract) Transfer(ctx contractapi.TransactionContextInterface, recipient string, amount int) error {
+
+	if amount <= 0 {
+		return fmt.Errorf("transfer amount must be a positive integer")
+	}
+
+	// Get ID of submitting client identity
+	clientID, err := ctx.GetClientIdentity().GetID()
+	if err != nil {
+		return fmt.Errorf("failed to get client id: %v", err)
+	}
+
+	clientCurrentBalanceBytes, err := ctx.GetStub().GetState(clientID)
+	if err != nil {
+		return fmt.Errorf("failed to read client account %s from world state: %v", clientID, err)
+	}
+
+	if clientCurrentBalanceBytes == nil {
+		return fmt.Errorf("client account %s has no balance", clientID)
+	}
+
+	clientCurrentBalance, _ := strconv.Atoi(string(clientCurrentBalanceBytes))
+
+	if clientCurrentBalance < amount {
```

Comment:
ERC20 allows transfer amount of 0 ("Transfers of 0 values MUST be treated as normal transfers and fire the Transfer event."), so this should allow even if clientCurrentBalance==0

Reply:
ok, I changed line 68 from `if amount <= 0` to `if amount < 0`